### ISSUE TITLE
HTTP low-level client

### ISSUE THREAD

**Issue by morenol** [typesense/typesense-rust#4]
## Description

reqwest, surf, isahc, other?

In the python library these are the http calls that are being used:https://github.com/typesense/typesense-python/blob/c0592c8936540397d436d92b634cc40005c42a69/typesense/api_call.py#L135

**Comment by emmanuelantony2000** [comment]
I was thinking of hyper for the async side of things, and expose the Executor, so that anyone could even use their own custom executors if they want to. Ofcourse we could provide tokio, async-std and smol as cargo features.

For the blocking side of things, [ureq](https://crates.io/crates/ureq) might be a good option.

**Comment by morenol** [comment]
That sounds great!

**Comment by emmanuelantony2000** [comment]
WASM support could be added using [`web-sys`](https://crates.io/crates/web-sys). More info [here](https://rustwasm.github.io/wasm-bindgen/examples/fetch.html).

**Comment by emmanuelantony2000** [comment]
I could start working on the lower level client.

**Comment by emmanuelantony2000** [comment]
An idea for the internal API that I have come up with:

---

```rust
/// A hyper client with pluggable Connectors and Executor
type HyperClient<C> = hyper::Client<C, hyper::Body>;

/// A WASM client
struct WasmClient {
    // ...
}

/// A Blocking Client
type BlockingClient = ureq::Agent;
```

---

I'll probably wrap them up rather than rexporting them as types, with custom functions with similar names to be used for different clients. Is this a good ides? Is there a better approach to tackle this.

**Comment by morenol** [comment]
I like the idea that is done in the elasticsearch client of Rust, where they have a HTTP low level struct called [Transport](https://docs.rs/elasticsearch/7.12.0-alpha.1/elasticsearch/http/transport/struct.Transport.html), and they create the [Client](https://docs.rs/elasticsearch/7.12.0-alpha.1/elasticsearch/struct.Elasticsearch.html) with that.

They have a [TransportBuilder](https://docs.rs/elasticsearch/7.12.0-alpha.1/elasticsearch/http/transport/struct.TransportBuilder.html) to create that.

What do you think about that approach?

**Comment by emmanuelantony2000** [comment]
Yeah this seems like a good option.

**Comment by morenol** [comment]
Closed by #8
